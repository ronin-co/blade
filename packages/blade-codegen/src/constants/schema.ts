import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';

import type { TypeNode } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { DML_QUERY_TYPES } from 'blade-compiler';

/**
 * A list of all model field types & their TypeScript type mapping.
 */
export const MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD = {
  blob: factory.createTypeReferenceNode(identifiers.compiler.storedObject),
  boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
  date: factory.createTypeReferenceNode(identifiers.primitive.date),
  json: factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.utils.jsonObject),
    factory.createTypeReferenceNode(identifiers.utils.jsonArray),
  ]),
  link: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
  number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
  string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
} satisfies Record<ModelField['type'], TypeNode>;

/**
 * A simple object mapping all DML query types to their human readable string.
 */
export const READABLE_DML_QUERY_TYPES = {
  add: 'Add',
  count: 'Count',
  get: 'Get',
  remove: 'Remove',
  set: 'Set',
  use: 'Get',
} satisfies Record<(typeof DML_QUERY_TYPES)[number] | 'use', string>;

/**
 * A list of all default field slugs for RONIN models.
 *
 * This is designed to match the properties as part of `ResultRecord`.
 */
export const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;
