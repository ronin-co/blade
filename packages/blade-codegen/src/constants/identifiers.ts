import { factory } from 'typescript';

import type { Identifier } from 'typescript';

/**
 * An identifier is the name of any type, interface, namespace, function, variable, etc.
 *
 * This can include any native utility types offered by TypeScript like `Partial`, `Record`, etc.
 *
 * Here we simply store a list of all identifiers used in the code generation package.
 */
export const identifiers = {
  blade: {
    module: {
      client: {
        hooks: factory.createIdentifier(JSON.stringify('blade/client/hooks')),
      },
      root: factory.createIdentifier(JSON.stringify('blade')),
      server: {
        hooks: factory.createIdentifier(JSON.stringify('blade/server/hooks')),
      },
      types: factory.createIdentifier(JSON.stringify('blade/types')),
    },
    reducedFunction: factory.createIdentifier('ReducedFunction'),
    resultRecord: factory.createIdentifier('ResultRecord'),
  },
  compiler: {
    combinedInstructions: factory.createIdentifier('CombinedInstructions'),
    dmlQueryType: {
      add: factory.createIdentifier('AddQuery'),
      count: factory.createIdentifier('CountQuery'),
      get: factory.createIdentifier('GetQuery'),
      remove: factory.createIdentifier('RemoveQuery'),
      set: factory.createIdentifier('SetQuery'),
      use: factory.createIdentifier('GetQuery'),
    },
    expression: factory.createIdentifier('Expression'),
    model: factory.createIdentifier('Model'),
    module: {
      root: factory.createIdentifier(JSON.stringify('blade-compiler')),
    },
    storedObject: factory.createIdentifier('StoredObject'),
  },
  namespace: {
    syntax: {
      fieldSlug: factory.createIdentifier('FieldSlug'),
      suffix: factory.createIdentifier('Syntax'),
      singular: factory.createIdentifier('Singular'),
      plural: factory.createIdentifier('Plural'),
    },
    utils: {
      afterQuery: factory.createIdentifier('AfterQuery'),
      afterQueryPromise: factory.createIdentifier('AfterQueryPromise'),
      beforeQuery: factory.createIdentifier('BeforeQuery'),
      beforeQueryPromise: factory.createIdentifier('BeforeQueryPromise'),
      name: factory.createIdentifier('Utils'),
      includingQuery: factory.createIdentifier('IncludingQuery'),
      includingQueryPromise: factory.createIdentifier('IncludingQueryPromise'),
      limitedToQuery: factory.createIdentifier('LimitedToQuery'),
      limitedToQueryPromise: factory.createIdentifier('LimitedToQueryPromise'),
      orderedByQuery: factory.createIdentifier('OrderedByQuery'),
      orderedByQueryPromise: factory.createIdentifier('OrderedByQueryPromise'),
      rootQueryCaller: factory.createIdentifier('RootQueryCaller'),
      rootQueryCallerPromise: factory.createIdentifier('RootQueryCallerPromise'),
      selectingQuery: factory.createIdentifier('SelectingQuery'),
      selectingQueryPromise: factory.createIdentifier('SelectingQueryPromise'),
      withQuery: factory.createIdentifier('WithQuery'),
      withQueryPromise: factory.createIdentifier('WithQueryPromise'),
    },
  },
  primitive: {
    array: factory.createIdentifier('Array'),
    date: factory.createIdentifier('Date'),
    partial: factory.createIdentifier('Partial'),
    promise: factory.createIdentifier('Promise'),
    record: factory.createIdentifier('Record'),
  },
  utils: {
    all: factory.createIdentifier('all'),
    jsonArray: factory.createIdentifier('JsonArray'),
    jsonObject: factory.createIdentifier('JsonObject'),
    jsonPrimitive: factory.createIdentifier('JsonPrimitive'),
    resolveSchema: factory.createIdentifier('ResolveSchema'),
  },
} satisfies Record<
  string,
  Record<string, Identifier | Record<string, Identifier | Record<string, Identifier>>>
>;

/**
 * A list of all generic names used in the `blade-codegen` package.
 *
 * Similar to `identifiers` but designed specifically for use as generic names.
 *
 * TODO(@nurodev): Remove unused identifiers
 */
export const typeArgumentIdentifiers = {
  default: factory.createIdentifier('T'),
  fields: factory.createIdentifier('F'),
  key: factory.createIdentifier('K'),
  queries: factory.createIdentifier('Q'),
  schema: factory.createIdentifier('S'),
  using: factory.createIdentifier('U'),
} satisfies Record<string, Identifier>;
