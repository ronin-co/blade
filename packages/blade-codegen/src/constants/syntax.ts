/**
 * A list of all inferred combined instruction properties for RONIN models.
 */
export const INFERRED_COMBINED_INSTRUCTION_PROPERTIES = [
  'after',
  'before',
  'including',
  'limitedTo',
  'orderedBy',
  'selecting',

  // TODO(@nurodev): Move out & only include `using` if the model includes any link fields.
  'using',
] satisfies Array<string>;
