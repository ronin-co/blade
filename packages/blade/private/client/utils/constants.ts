export const IS_CLIENT_DEV = import.meta.env.BLADE_ENV === 'development';
