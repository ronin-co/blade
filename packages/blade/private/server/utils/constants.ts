import { DDL_QUERY_TYPES, DML_QUERY_TYPES_WRITE } from 'blade-compiler';

export const VERBOSE_LOGGING = import.meta.env.__BLADE_DEBUG_LEVEL === 'verbose';
export const IS_SERVER_DEV = import.meta.env.BLADE_ENV === 'development';

/** A list of all query types that update the database. */
export const WRITE_QUERY_TYPES = [
  ...DML_QUERY_TYPES_WRITE,
  ...DDL_QUERY_TYPES.filter((item) => item !== 'list'),
];
