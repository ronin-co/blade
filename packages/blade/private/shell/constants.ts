import path from 'node:path';
import chalk from 'chalk';
import gradient from 'gradient-string';

export const pagesDirectory = path.resolve(process.cwd(), 'pages');
export const componentsDirectory = path.resolve(process.cwd(), 'components');
export const triggersDirectory = path.resolve(process.cwd(), 'triggers');
export const publicDirectory = path.resolve(process.cwd(), 'public');

export const frameworkDirectory = path.join(__dirname, '..', '..', '..');

export const directoriesToParse = {
  pages: pagesDirectory,
  triggers: triggersDirectory,
};

export const outputDirectory = path.resolve(process.cwd(), '.blade');

// The path at which people can define a custom Hono app that Blade will mount.
export const routerInputFile = path.join(process.cwd(), 'router.ts');

export const styleInputFile = path.join(process.cwd(), 'styles.css');
export const clientInputFile = require.resolve('./private/client/index.js');

const serverInputFile = require.resolve(
  './private/server/worker/providers/edge-worker.js',
);
export const serverInputFolder = path.dirname(serverInputFile);

export const loggingPrefixes = {
  info: `${chalk.bold(gradient(['#473b7b', '#3584a7', '#30d2be'])('BLADE'))} `,
  error: `${chalk.bold(gradient(['#930024', '#d4143e'])('ERROR'))}  `,
};

export const defaultDeploymentProvider = 'edge-worker';
