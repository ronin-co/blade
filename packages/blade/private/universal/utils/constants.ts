export const CLIENT_ASSET_PREFIX = 'client';
export const PUBLIC_ASSET_PREFIX = 'public';

export const DEFAULT_PAGE_PATH = 'blade-default-pages';

export const CUSTOM_HEADERS = {
  bundleId: 'X-Bundle-Id',
  subscribe: 'X-Subscribe',
} as const;

// 365 days
export const DEFAULT_COOKIE_MAX_AGE = 31536000;
