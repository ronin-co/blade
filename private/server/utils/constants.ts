export const VERBOSE_LOGGING = import.meta.env.__BLADE_DEBUG_LEVEL === 'verbose';

export const NOT_FOUND_PAGE = '404';
export const EXCEPTION_PAGE = '500';

/**
 * These security headers are required to ensure a basic level of security.
 *
 * The defaults here are based on the recommended headers set in the Hono
 * `secureHeaders` middleware.
 *
 * @see https://github.com/honojs/hono/blob/bd36ad10fccf337408d732372dbdfd495c654cf2/src/middleware/secure-headers/index.ts#L67-L80
 */
export const SECURITY_HEADERS = {
  'Cross-Origin-Opener-Policy': 'same-origin',
  'Cross-Origin-Resource-Policy': 'same-origin',
  'Origin-Agent-Cluster': '?1',
  'Referrer-Policy': 'no-referrer',
  'Strict-Transport-Security': 'max-age=15552000; includeSubDomains',
  'X-Content-Type-Options': 'nosniff',
  'X-DNS-Prefetch-Control': 'off',
  'X-Download-Options': 'noopen',
  'X-Frame-Options': 'SAMEORIGIN',
  'X-Permitted-Cross-Domain-Policies': 'none',
  'X-XSS-Protection': '0',
};
