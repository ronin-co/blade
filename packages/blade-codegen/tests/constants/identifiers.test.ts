import { expect, test } from 'bun:test';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';

test('identifiers', () => {
  expect(identifiers).toMatchSnapshot();
});

test('generic identifiers', () => {
  expect(typeArgumentIdentifiers).toMatchSnapshot();
});
