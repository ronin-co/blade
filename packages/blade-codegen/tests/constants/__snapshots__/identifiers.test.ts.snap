// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`identifiers 1`] = `
{
  "blade": {
    "module": {
      "client": {
        "hooks": IdentifierObject {
          "emitNode": undefined,
          "end": -1,
          "escapedText": ""blade/client/hooks"",
          "flags": 16,
          "flowNode": undefined,
          "id": 0,
          "jsDoc": undefined,
          "kind": 80,
          "parent": undefined,
          "pos": -1,
          "symbol": undefined,
          "transformFlags": 0,
        },
      },
      "root": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": ""blade"",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "server": {
        "hooks": IdentifierObject {
          "emitNode": undefined,
          "end": -1,
          "escapedText": ""blade/server/hooks"",
          "flags": 16,
          "flowNode": undefined,
          "id": 0,
          "jsDoc": undefined,
          "kind": 80,
          "parent": undefined,
          "pos": -1,
          "symbol": undefined,
          "transformFlags": 0,
        },
      },
      "types": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": ""blade/types"",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
    },
    "reducedFunction": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "ReducedFunction",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "resultRecord": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "ResultRecord",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
  },
  "compiler": {
    "combinedInstructions": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "CombinedInstructions",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "dmlQueryType": {
      "add": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "AddQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "count": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "CountQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "get": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "GetQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "remove": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "RemoveQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "set": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "SetQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "use": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "GetQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
    },
    "expression": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Expression",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "model": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Model",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "module": {
      "root": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": ""blade-compiler"",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
    },
    "storedObject": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "StoredObject",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
  },
  "namespace": {
    "syntax": {
      "fieldSlug": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "FieldSlug",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "plural": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "Plural",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "singular": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "Singular",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "suffix": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "Syntax",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
    },
    "utils": {
      "afterQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "AfterQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "afterQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "AfterQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "beforeQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "BeforeQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "beforeQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "BeforeQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "includingQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "IncludingQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "includingQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "IncludingQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "limitedToQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "LimitedToQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "limitedToQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "LimitedToQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "name": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "Utils",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "orderedByQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "OrderedByQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "orderedByQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "OrderedByQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "rootQueryCaller": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "RootQueryCaller",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "rootQueryCallerPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "RootQueryCallerPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "selectingQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "SelectingQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "selectingQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "SelectingQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "toQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "ToQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "toQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "ToQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "usingQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "UsingQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "usingQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "UsingQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "withQuery": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "WithQuery",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
      "withQueryPromise": IdentifierObject {
        "emitNode": undefined,
        "end": -1,
        "escapedText": "WithQueryPromise",
        "flags": 16,
        "flowNode": undefined,
        "id": 0,
        "jsDoc": undefined,
        "kind": 80,
        "parent": undefined,
        "pos": -1,
        "symbol": undefined,
        "transformFlags": 0,
      },
    },
  },
  "primitive": {
    "array": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Array",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "date": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Date",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "partial": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Partial",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "promise": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Promise",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "record": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "Record",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
  },
  "utils": {
    "all": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "all",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "jsonArray": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "JsonArray",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "jsonObject": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "JsonObject",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "jsonPrimitive": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "JsonPrimitive",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
    "resolveSchema": IdentifierObject {
      "emitNode": undefined,
      "end": -1,
      "escapedText": "ResolveSchema",
      "flags": 16,
      "flowNode": undefined,
      "id": 0,
      "jsDoc": undefined,
      "kind": 80,
      "parent": undefined,
      "pos": -1,
      "symbol": undefined,
      "transformFlags": 0,
    },
  },
}
`;

exports[`generic identifiers 1`] = `
{
  "default": IdentifierObject {
    "emitNode": undefined,
    "end": -1,
    "escapedText": "T",
    "flags": 16,
    "flowNode": undefined,
    "id": 0,
    "jsDoc": undefined,
    "kind": 80,
    "parent": undefined,
    "pos": -1,
    "symbol": undefined,
    "transformFlags": 0,
  },
  "fields": IdentifierObject {
    "emitNode": undefined,
    "end": -1,
    "escapedText": "F",
    "flags": 16,
    "flowNode": undefined,
    "id": 0,
    "jsDoc": undefined,
    "kind": 80,
    "parent": undefined,
    "pos": -1,
    "symbol": undefined,
    "transformFlags": 0,
  },
  "key": IdentifierObject {
    "emitNode": undefined,
    "end": -1,
    "escapedText": "K",
    "flags": 16,
    "flowNode": undefined,
    "id": 0,
    "jsDoc": undefined,
    "kind": 80,
    "parent": undefined,
    "pos": -1,
    "symbol": undefined,
    "transformFlags": 0,
  },
  "schema": IdentifierObject {
    "emitNode": undefined,
    "end": -1,
    "escapedText": "S",
    "flags": 16,
    "flowNode": undefined,
    "id": 0,
    "jsDoc": undefined,
    "kind": 80,
    "parent": undefined,
    "pos": -1,
    "symbol": undefined,
    "transformFlags": 0,
  },
  "using": IdentifierObject {
    "emitNode": undefined,
    "end": -1,
    "escapedText": "U",
    "flags": 16,
    "flowNode": undefined,
    "id": 0,
    "jsDoc": undefined,
    "kind": 80,
    "parent": undefined,
    "pos": -1,
    "symbol": undefined,
    "transformFlags": 0,
  },
}
`;
